README.md
![VERSION](https://img.shields.io/github/v/tag/PascalCoin/PascalCoin?sort=semver&style=for-the-badge)
![License](https://img.shields.io/github/license/PascalCoin/PascalCoin.svg?style=for-the-badge)
![DISCORD](https://img.shields.io/discord/383064643482025984.svg?label=DISCORD%20COMMUNITY&style=for-the-badge)
![REDDIT](https://img.shields.io/reddit/subreddit-subscribers/pascalcoin.svg?style=for-the-badge)
<p align="center">
<br /><br />
<img width="30%" src="resources/logo.svg" />
</p>
<br /><br />


```
Copyright (c) 2016-2019 Pascal developers based on original Albert Molina source code
  
THIS IS EXPERIMENTAL SOFTWARE. Use it for educational purposes only.
```

# Pascal - A One of a kind Cryptocurrency

> Pascal is an instant, zero-fee, infinitely scalable, and decentralized cryptocurrency with advanced privacy and smart contract capabilities. It's the world's first blockchain independent of historical operations.

> Pascal possesses unlimited potential.

## License

The source code is distributed under the `MIT software license`, see the accompanying [LICENSE](LICENSE) file or visit http://www.opensource.org/licenses/mit-license.php.

This product includes software developed by the OpenSSL Project and Denis Grinyuk (https://github.com/Arvur/OpenSSL-Delphi) and some cryptographic functions inspirated in code written by Ladar Levison and Marco Ferrante.

Go to https://www.pascalcoin.org to find out more.


## Releases

 - [**MAINNET v4**](https://github.com/PascalCoin/PascalCoin/releases/tag/4.1)
 - [**TESTNET v5**](https://github.com/PascalCoin/PascalCoin/releases/tag/TESTNET_5.beta.1)

## Community

In case you have questions or just want to get in touch with like-minded Pascal users.

 - Listen to our twitter on: https://twitter.com/PascalCoin
 - Chat with us on discord: https://discord.gg/9YYZyw
 - Join our subreddit on: https://discord.gg/9YYZyw

## Develop with Pascal

| Language      | Project Name | Type    | Version support | Link |
| ------------- | ------------ | ------- | --------------- | ---- |
| JavaScript    | SBX          | Library | 4    | [https://github.com/techworker/sbx](SBX) |
| Dart          | PascalDart   | Library | 4    | [https://github.com/appditto/pascaldart](PascalDart) |


## Changelog

View the changelog [here](CHANGELOG.md)


CHANGELOG
## History:  

### Current Build (Pending release date)
- Upgrade to Protocol 5 (Hard fork)
- Implementation of PIP-0032 (Atomic Swaps) -> https://github.com/PascalCoin/PascalCoin/blob/master/PIP/PIP-0032.md
- Implementation of PIP-0030 (Safebox root) -> https://github.com/PascalCoin/PascalCoin/blob/master/PIP/PIP-0030.md
- Implementation of PIP-0029 (Account Seals) -> https://github.com/PascalCoin/PascalCoin/blob/master/PIP/PIP-0029.md
- Implementation of PIP-0024 (Account Data) -> https://github.com/PascalCoin/PascalCoin/blob/master/PIP/PIP-0024.md
- Implementation of PIP-0033 (OpData JOSN-RPC calls) -> https://github.com/PascalCoin/PascalCoin/blob/master/PIP/PIP-0033.md
- Partial implementation of PIP-0012 (Recover Accounts option after 4 years) -> https://github.com/PascalCoin/PascalCoin/blob/master/PIP/PIP-0012.md
- Updated "OP_DATA" operation: (PIP-0016)
  - New digest hash value for OP_DATA ( PIP-0016 ) on Protocol 5
  - Added "id" field (GUID/UUID type as described on PIP-0016), was missing on V4, added on V5
- Updated "OP_CHANGE_ACCOUNT_INFO" and "OP_MULTIOPERATION" to allow Account.Data as described on PIP-0024
  - Added "new_data" field allowing update Account.Data field (0..32 bytes)
  - Updated digest hash value adding "new_data" field  
- Hardcoded RandomHash digest/hash values for quick speed safebox check on fresh installation
- JSON-RPC changes:  
  - Updated "listaccountforsale" method to allow ATOMIC SWAPS (PIP-0032)
    - Added "type" to discrimine between kind of listing. Available values are:
      - "public_sale"
      - "private_sale": Need to provide a valid "new_enc_pubkey" or "new_b58_pubkey"
      - "atomic_account_swap": Need to provide a valid "new_enc_pubkey" or "new_b58_pubkey" and the unlocking param "enc_hash_lock"
      - "atomic_coin_swap": Need to provide a valid "enc_hash_lock"
      - If no "type" is defined, will automatically select between "public_sale" or "private_sale"
    - Added "enc_hash_lock" (HEXASTRING) that must be exactly a 32 bytes value (stored as 64 bytes because is HexaString)
  - Updated "changeaccountinfo" and "signchangeaccountinfo" calls to allow add "new_data" field for change Account.Data value (PIP-0024)
    - New param "new_data" (HEXASTRING) if provided will change Account Data info. Limited from 0 to 32 bytes.
  - New method "senddata" as described on PIP-0033 returning an "Operation Object"
  - New method "signdata" as described on PIP-0033 returning a "Raw Operations Object"
  - New method "finddataoperations" as described on PIP-0033 returning an ARRAY of "Raw Operations Object"
    - News params added to original PIP
      - "depth" (Integer, 1000 by default) : Will search backward in blocks a maximum of "depth" blocks
      - "startblock" (Integer, optional) : If defined, will start at a specified block, otherwise will start at last "sender" or "target" updated block
  - Updated "Account Object" return values:
    - "state": Can return "normal", "listed", "account_swap", "coin_swap"
    - "hashed_secret" : (HEXASTRING) will contain the SHA256( SECRET ) value that must match Payload received data for Atomic Swaps (only when "state" in "account_swap" or "coin_swap")
    - "amount_to_swap" : (PASCURRENCY) amount that will be transferred to counterparty account on ATOMIC COIN SWAP ("receiver_swap_account")
    - "receiver_swap_account": (Integer) Counterpaty account that will receive "amount_to_swap" on ATOMIC COIN SWAP
    - "data" : (HEXASTRING) will return the Account Data stored with PIP-0024
    - "seal" : (HEXASTRING) will return the Account Seal stored with PIP-0029
  - Updated "Operation Object" return values:
    - "senders" : ARRAY
      - "data" : OBJECT will store OP_DATA information when operation is OP_DATA type as described on PIP-0016
        - "id" : (String) String representation of GUID/UUID as "00000000-0000-0000-0000-000000000000" that stores 16 bytes
        - "sequence" : (Integer)
        - "type" : (Integer)
    - "changers" : ARRAY
      - "new_data" : (HEXASTRING) : If "data" is changed on "account"
      - "changes" : (String) Description of changes type made
  - Updated "Multi Operation Object" values:
    - "changers" : ARRAY
      - "new_data" : (HEXASTRING) : If "data" is changed on "account"

TODO  
- TODO: RPC calls for PIP-0029
- TODO: RPC calls for PIP-0030
- TODO: RPC calls for PIP-0016

### Build 4.1.0.0 - 2019-07-24
- Hardcoded RandomHash digest/hash values for quick speed safebox check on fresh installation
  - GUI wallets will load "HardcodedRH_75800.randomhash" file (at exe folder) and preload a randomhash digest/hash values for quick first time synchronization
- Fixed bug caused by CT_NetOp_GetSafeBox too quickly on old versions (4.0.2 and lower)
  - Added a delay of at least 1 second per call when peer node is protocol version lower than version 9
- Fixed bug #187 found by Isaac Cook (icook)
- Improved memory management on NetProtection (auto clean memory)

### Build 4.0.3.1 - 2019-04-12
- Fixed core bug #182 in RPC calls

### Build 4.0.3 - 2019-04-10
- Improvements:
  - Up to +1600 operations per second using a domestic computer (4 CPU)
  - Download and check new Safebox in less than 5 minutes
- TESTNET:
  - Added functions to automatically receive new accounts for new installed fresh nodes
- Major refactoring: 
  - Change TRawBytes from "AnsiString" to "TBytes" type
  - Do not use AnsiString type for simple string types, use String instead
  - Usage of TList<T> type
  - Allow compilation using Delphi 10.2 and multidevice (Firemonkey)
- Core features
  - Added CryptoLib4Pascal library that will allow to replace OpenSSL in a future for a Pascal pure native cryptographic suite (for testing purposes at this moment, not actived by default)
  - Important log reduction
  - Adding multithreading validation
    - Multithreading validating Signatures
    - Multithreading validating Blocks headers (TOperationBlock)
- Fixed core bugs:

src/core/UPCHardcodedRandomHashTable.pas
unit UPCHardcodedRandomHashTable;

{ Copyright (c) 2019 by Albert Molina

  Distributed under the MIT software license, see the accompanying file LICENSE
  or visit http://www.opensource.org/licenses/mit-license.php.

  This unit is a part of the PascalCoin Project, an infinitely scalable
  cryptocurrency. Find us here:
  Web: https://www.pascalcoin.org
  Source: https://github.com/PascalCoin/PascalCoin

  If you like it, consider a donation using Bitcoin:
  16K3HCZRhFUtM8GdWRcfKeaa6KsuyxZaYk

  THIS LICENSE HEADER MUST NOT BE REMOVED.
}

{$IFDEF FPC}
  {$MODE Delphi}
{$ENDIF}

{$I config.inc}

interface

uses
  SysUtils, Classes,
  UCrypto, UBaseTypes, UPCDataTypes, UConst,
  {$IFNDEF FPC}System.Generics.Collections{$ELSE}Generics.Collections{$ENDIF};


type
  TPCHardcodedRandomHashTable = Class
    {
     This object will store an ordered list of pairs <Digest><RandomHash>
     ordered by <Digest> for quick searching.
     Is usefull to store Hardcoded values of RandomHash (as described
     on PIP-0009)
     https://github.com/PascalCoin/PascalCoin/blob/master/PIP/PIP-0009.md

     For security reasons, this object data will be stored in a secured Stream
     (For example a file) where last value of the stream is a SHA256 hash
     of the previous content.

     The "LoadFromStream" function will return TRUE only if the SHA256 hash
     stored in the Stream matches the loaded data, preventing corruption

     Also, an external App can hardcode this SHA256 value (obtained
     in a call to GetHardcodedSha256) for securize this usage

     }
    Type
    TRow = Record
      RandomHashValue : T32Bytes;
      DigestValue : TRawBytes;
    End;
  private
    FList : TList<TRow>;
    FHardcodedSha256: TRawBytes;
    function Find(const ADigest : TRawBytes; var AIndex : Integer) : Boolean;
    function GetRandomHash(AIndex : Integer) : TRawBytes;
    function GetRow(AIndex : Integer) : TRow;
    procedure Clear;
    procedure SaveToStreamWithoutHardcodedSha256(AStream : TStream);
  public
    constructor Create;
    destructor Destroy; override;
    function LoadFromStream(AStream : TStream; var AHardcodedSha256 : TRawBytes) : Boolean;
    procedure SaveToStream(AStream : TStream);
    procedure AddRandomHash(const ARandomHash, ADigest : TRawBytes; ACheckIntegrity : Boolean);
    function Count : Integer;
    function FindRandomHashByDigest(const ADigest : TRawBytes; var ARandomHash : TRawBytes) : Boolean;
    function GetHardcodedSha256 : TRawBytes;  // Will return a SHA256 of the object content for hardcode external apps
    procedure CopyFrom(ASource : TPCHardcodedRandomHashTable);
  End;

implementation

uses UAccounts;

const
  CT_MagicValue_Stream_Header = 'TPCHardcodedRandomHashTable';
  CT_MagicValue_Stream_Version = 1;

{ TPCHardcodedRandomHashTable }

procedure TPCHardcodedRandomHashTable.AddRandomHash(const ARandomHash, ADigest: TRawBytes; ACheckIntegrity: Boolean);
var i,j, LMemPos_0, LMemPos_1 : Integer;
  LRaw : TRawBytes;
  LRow : TRow;
  LRowPtr : Pointer;
  LRowPosition : Int64;
begin
  if ACheckIntegrity then begin
    LRaw := TCrypto.DoRandomHash(ADigest);
    if TBaseType.BinStrComp(LRaw,ARandomHash)<>0 then raise Exception.Create('RandomHash(Digest) & Provided Hash value does not match');
  end;
  if Length(ARandomHash)<>32 then raise Exception.Create('Invalid RandomHash length<>32');

  if Find(ADigest,i) then begin
    if TBaseType.BinStrComp(GetRandomHash(i),ARandomHash)<>0 then raise Exception.Create('Digest found with another RandomHash value');
  end else begin
    FHardcodedSha256 := Nil;

    LRow.RandomHashValue := TBaseType.To32Bytes(ARandomHash);
    LRow.DigestValue := ADigest;
    FList.Insert(i,LRow);
  end;
end;

procedure TPCHardcodedRandomHashTable.Clear;
begin
  FList.Clear;
  FHardcodedSha256 := Nil;
end;

procedure TPCHardcodedRandomHashTable.CopyFrom(ASource: TPCHardcodedRandomHashTable);
var i, iIndex : Integer;
begin
  if ASource=Self then Exit;
  for i := 0 to ASource.Count-1 do begin
    if Not Find(ASource.FList[i].DigestValue,iIndex) then begin
      FList.Insert(iIndex,ASource.FList[i]);
      FHardcodedSha256 := Nil;
    end;
  end;
end;

function TPCHardcodedRandomHashTable.Count: Integer;
begin
  Result := FList.Count;
end;

constructor TPCHardcodedRandomHashTable.Create;
begin
  FList := TList<TRow>.Create;
  FHardcodedSha256 := Nil;
end;

destructor TPCHardcodedRandomHashTable.Destroy;
begin
  Clear;
  FreeAndNil(FList);
  inherited;
end;

function TPCHardcodedRandomHashTable.Find(const ADigest : TRawBytes; var AIndex : Integer) : Boolean;
var L, H, I: Integer;
  C : Int64;
  LPtr : PByte;
  LRowDigest : TRawBytes;
begin
  Result := False;
  L := 0;
  H := Count-1;
  while L <= H do
  begin
    I := (L + H) shr 1;

    LRowDigest := GetRow(I).DigestValue;

    C := TBaseType.BinStrComp(LRowDigest,ADigest);
    if C < 0 then L := I + 1 else
    begin
      H := I - 1;
      if C = 0 then
      begin
        Result := True;
        L := I;
      end;
    end;
  end;
  AIndex := L;
end;

function TPCHardcodedRandomHashTable.FindRandomHashByDigest(const ADigest: TRawBytes; var ARandomHash: TRawBytes): Boolean;
var i : Integer;
begin
  if Find(ADigest,i) then begin
    ARandomHash := TBaseType.T32BytesToRawBytes( FList.Items[i].RandomHashValue );
    Result := True;
  end else begin
    Result := False;
    SetLength(ARandomHash,0);
  end;
end;

function TPCHardcodedRandomHashTable.GetHardcodedSha256: TRawBytes;
var LMemStream : TMemoryStream;
begin
  if Length(FHardcodedSha256)<>32 then begin
    LMemStream := TMemoryStream.Create;
    try
      SaveToStreamWithoutHardcodedSha256(LMemStream);
      FHardcodedSha256 := TCrypto.DoSha256( PAnsiChar(LMemStream.Memory), LMemStream.Size );
    finally
      LMemStream.Free;
    end;
  end;
  Result := FHardcodedSha256;
end;

function TPCHardcodedRandomHashTable.GetRandomHash(AIndex: Integer): TRawBytes;
begin
  Result := TBaseType.ToRawBytes( GetRow(AIndex).RandomHashValue );
end;

function TPCHardcodedRandomHashTable.GetRow(AIndex: Integer): TRow;
begin
  Result := FList.Items[AIndex];
end;

function TPCHardcodedRandomHashTable.LoadFromStream(AStream: TStream; var AHardcodedSha256 : TRawBytes): Boolean;
var i : Integer;
  LRaw, LRaw2 : TRawBytes;
  LTotalRows : UInt32;
  LVersion : Word;
begin
  Clear;
  TStreamOp.ReadAnsiString(AStream,LRaw);
  LRaw2.FromString(CT_MagicValue_Stream_Header);
  Result := False;
  if (Not TBaseType.Equals(LRaw,LRaw2)) then begin
    Exit(False);
  end;
  AStream.Read(LVersion,2);
  if Not LVersion=CT_MagicValue_Stream_Version then Exit(False);

  AStream.Read(LTotalRows,4);
  for i := 1 to LTotalRows do begin
    if TStreamOp.ReadAnsiString(AStream,LRaw)<0 then Exit(False); // RandomHash value
    if TStreamOp.ReadAnsiString(AStream,LRaw2)<0 then Exit(False);// Digest value
    AddRandomHash(LRaw,LRaw2,False);
  end;

  // Last value must be a
  if TStreamOp.ReadAnsiString(AStream,AHardcodedSha256)<0 then Exit(False);
  if TBaseType.BinStrComp(AHardcodedSha256,GetHardcodedSha256)<>0 then Exit(False);

  Result := True;
end;

procedure TPCHardcodedRandomHashTable.SaveToStream(AStream: TStream);
begin
  SaveToStreamWithoutHardcodedSha256(AStream);

  // Last value to the Stream must be
  TStreamOp.WriteAnsiString(AStream,GetHardcodedSha256);
end;

procedure TPCHardcodedRandomHashTable.SaveToStreamWithoutHardcodedSha256(AStream: TStream);
var LVersion : Word;
  LRaw : TRawBytes;
  LTotalRows : UInt32;
  i : Integer;
begin
  LRaw.FromString(CT_MagicValue_Stream_Header);
  TStreamOp.WriteAnsiString(AStream,LRaw);
  LVersion := CT_MagicValue_Stream_Version;
  AStream.Write(LVersion,2);
  //
  LTotalRows := Count;
  AStream.Write(LTotalRows,4);
  for i := 0 to Count-1 do begin
    LRaw := TBaseType.ToRawBytes(FList[i].RandomHashValue);
    TStreamOp.WriteAnsiString(AStream,LRaw);
    TStreamOp.WriteAnsiString(AStream,FList[i].DigestValue);
  end;
end;

end.


src/core/UConst.pas
unit UConst;

{ Copyright (c) 2016 by Albert Molina

  Distributed under the MIT software license, see the accompanying file LICENSE
  or visit http://www.opensource.org/licenses/mit-license.php.

  This unit is a part of the PascalCoin Project, an infinitely scalable
  cryptocurrency. Find us here:
  Web: https://www.pascalcoin.org
  Source: https://github.com/PascalCoin/PascalCoin

  If you like it, consider a donation using Bitcoin:
  16K3HCZRhFUtM8GdWRcfKeaa6KsuyxZaYk

  THIS LICENSE HEADER MUST NOT BE REMOVED.
}

{$IFDEF FPC}
  {$MODE Delphi}
{$ENDIF}

interface

{$I config.inc}

{$IFNDEF FPC}
  // See http://wiki.freepascal.org/Code_Conversion_Guide
type
  PtrInt = NativeInt;
  PtrUInt = NativeUInt;
  TThreadID = NativeUInt;
{$ENDIF}

Const
  CT_Genesis_Magic_String_For_Old_Block_Hash :
    String =
    '(c) Albert Molina - Genesis block at same time than BitCoin Block 424720 Hash 000000000000000001cc41ff7846264718ef0a15f97f532a98277bd5f6820b89';

  CT_Zero_Block_Proof_of_work_in_Hexa =
    {$IFDEF PRODUCTION}'00000003A29C32E84A539ADE24397D41D30116A6FAFEC17B7D9CED68A4238C92'{$ELSE}{$IFDEF TESTNET}''{$ELSE}{$ENDIF}{$ENDIF};


  CT_NetServer_Port = {$IFDEF PRODUCTION}4004{$ELSE}{$IFDEF TESTNET}4104{$ELSE}{$ENDIF}{$ENDIF};
  CT_JSONRPCMinerServer_Port = {$IFDEF PRODUCTION}4009{$ELSE}{$IFDEF TESTNET}4109{$ELSE}{$ENDIF}{$ENDIF};
  CT_JSONRPC_Port = {$IFDEF PRODUCTION}4003{$ELSE}{$IFDEF TESTNET}4103{$ELSE}{$ENDIF}{$ENDIF};
  CT_AccountsPerBlock = 5;

  CT_NewLineSecondsAvg: Cardinal = {$IFDEF PRODUCTION}300{$ELSE}{$IFDEF TESTNET}30{$ELSE}{$ENDIF}{$ENDIF};
    // 60*5=300 seconds -> 5 minutes avg
    //   -> 1 day = 86400 seconds -> 1 year = 31536000 seconds (aprox)
    //   Each year = 105120 new blocks (aprox)
    //   -> *5 accounts per block = 525600 new accounts each year (aprox)

  CT_FirstReward: UInt64 = 1000000; // 4 decimals... First reward = 100,0000
  CT_MinReward: UInt64 = 10000; // 4 decimals... Min reward = 1,0000
  CT_NewLineRewardDecrease: Cardinal = {$IFDEF PRODUCTION}210240{$ELSE}{$IFDEF TESTNET}20000{$ENDIF}{$ENDIF};

  CT_WaitNewBlocksBeforeTransaction = 100;

  CT_RecoverFoundsWaitInactiveCount = 420480;  // After 4 years... if an account has no operations, money will be a reward for a miner!
  CT_MaxFutureBlocksLockedAccount = 105120; // Maximum future blocks an account can be locked

  CT_MaxTransactionAmount = 1000000000000; // ... can be deleted
  CT_MaxTransactionFee = 100000000;
  CT_MaxWalletAmount = 10000000000000; // ... can be deleted
  //
  CT_MinCompactTarget_v1: Cardinal = {$IFDEF PRODUCTION}$19000000{$ELSE}{$IFDEF TESTNET}$17000000{$ELSE}{$ENDIF}{$ENDIF}; // First compact target of block 0
  CT_MinCompactTarget_v4: Cardinal = // Minimum compact target of block if using Protocol 4 or higher
    {$IFDEF ACTIVATE_RANDOMHASH_V4}
    {$IFDEF PRODUCTION}$12000000{$ELSE}{$IFDEF TESTNET}$08000000{$ELSE}{$ENDIF}{$ENDIF}
    {$ELSE}CT_MinCompactTarget_v1{$ENDIF};

  {$IFDEF ACTIVATE_RANDOMHASH_V4}
  CT_CompactTarget_Reset_v4: Cardinal = // First compact target of block if using Protocol 4 and RandomHash is Active
    {$IFDEF PRODUCTION}$16000000{$ELSE}$08000000{$ENDIF};
  {$ENDIF}


  CT_CalcNewTargetBlocksAverage: Cardinal = 100;
  CT_CalcNewTargetLimitChange_SPLIT = 10;

  CT_MaxAccount : Cardinal = $FFFFFFFF;
  CT_MaxBlock : Cardinal = $FFFFFFFF;

  CT_MaxPayloadSize = 255; // Max payload size in bytes
  CT_MaxFutureBlockTimestampOffset = 15;
  CT_MinNodesToCalcNAT = 4;

  CT_MinServersConnected = 2;
  CT_MaxServersConnected = 5;

  CT_MaxResendMemPoolOperations = 50000;

  CT_MaxClientsConnected = {$IFDEF FPC}140{$ELSE}80{$ENDIF};

  CT_BankToDiskEveryNBlocks = {$IFDEF PRODUCTION}100{$ELSE}100{$ENDIF}; // Build 1.5 changed from 500 to 100;

  CT_NID_secp256k1 = 714;
  CT_NID_secp384r1 = 715;
  CT_NID_sect283k1 = 729;
  CT_NID_secp521r1 = 716;

  CT_Default_EC_OpenSSL_NID = CT_NID_secp256k1;

  CT_AccountInfo_ForSale = 1000;

  CT_PROTOCOL_1 = 1;
  CT_PROTOCOL_2 = 2;
  CT_PROTOCOL_3 = 3;
  CT_PROTOCOL_4 = 4;
  CT_BUILD_PROTOCOL = CT_PROTOCOL_4;

  CT_BlockChain_Protocol_Available: Word = 4; // Protocol 4 flag
  CT_Protocol_Upgrade_v2_MinBlock = {$IFDEF PRODUCTION}115000{$ELSE}50{$ENDIF};
  CT_Protocol_Upgrade_v3_MinBlock = {$IFDEF PRODUCTION}210000{$ELSE}250{$ENDIF};
  CT_Protocol_Upgrade_v4_MinBlock = {$IFDEF PRODUCTION}260000{$ELSE}400{$ENDIF};


  CT_MagicNetIdentification = {$IFDEF PRODUCTION}$0A043580{$ELSE}$04000000{$ENDIF}; // Unix timestamp 168048000 ... It's Albert birthdate!

  CT_NetProtocol_Version: Word = $0009; // Version 4.0.2 Will allow only net protocol 9
  // IMPORTANT NOTE!!!
  // NetProtocol_Available MUST BE always >= NetProtocol_version
  CT_NetProtocol_Available: Word = {$IFDEF PRODUCTION}$0009{$ELSE}$0009{$ENDIF};  // Version 4.0.0 will start accepting protocol 8 but 4.0.1 will accept 9 due to 4.0.0 bug

  CT_MaxAccountOperationsPerBlockWithoutFee = 1;

  CT_SafeBoxBankVersion : Word = 3; // Protocol 2 upgraded safebox version from 2 to 3

  CT_MagicIdentificator: String = {$IFDEF PRODUCTION}'PascalCoin'{$ELSE}'PascalCoinTESTNET'{$ENDIF}; //

  CT_PseudoOp_Reward = $0;
  // Value of Operations type in Protocol 1
  CT_Op_Transaction = $01;
  CT_Op_Changekey = $02;
  CT_Op_Recover = $03;
  // Protocol 2 new operations
  CT_Op_ListAccountForSale = $04;
  CT_Op_DelistAccount = $05;
  CT_Op_BuyAccount = $06;
  CT_Op_ChangeKeySigned = $07;
  CT_Op_ChangeAccountInfo = $08;
  // Protocol 3 new operations
  CT_Op_MultiOperation = $09;  // PIP-0017
  // Protocol 4 new operations
  CT_Op_Data = $0A;            // PIP-0016

  CT_Protocol_v3_PIP11_Percent = 20; // PIP-0011 20% Percent proposed and voted by PIP-0011

  CT_Hardcoded_RandomHash_Table_Filename = 'HardcodedRH_75800.randomhash';
  CT_Hardcoded_RandomHash_Table_HASH = '0A56291E8368AC855227B67A2F6CBEEFD2DB4CAE3CB8A473A7F6663C63368D0E';

  CT_PseudoOpSubtype_Miner                = 1;
  CT_PseudoOpSubtype_Developer            = 2;

  CT_OpSubtype_TransactionSender          = 11;
  CT_OpSubtype_TransactionReceiver        = 12;
  CT_OpSubtype_BuyTransactionBuyer        = 13;
  CT_OpSubtype_BuyTransactionTarget       = 14;
  CT_OpSubtype_BuyTransactionSeller       = 15;
  CT_OpSubtype_ChangeKey                  = 21;
  CT_OpSubtype_Recover                    = 31;
  CT_OpSubtype_ListAccountForPublicSale   = 41;
  CT_OpSubtype_ListAccountForPrivateSale  = 42;
  CT_OpSubtype_DelistAccount              = 51;
  CT_OpSubtype_BuyAccountBuyer            = 61;
  CT_OpSubtype_BuyAccountTarget           = 62;
  CT_OpSubtype_BuyAccountSeller           = 63;
  CT_OpSubtype_ChangeKeySigned            = 71;
  CT_OpSubtype_ChangeAccountInfo          = 81;
  CT_OpSubtype_MultiOperation_Global      = 91;
  CT_OpSubtype_MultiOperation_AccountInfo = 92;
  CT_OpSubtype_Data_GlobalInfo            = 101;
  CT_OpSubtype_Data_Sender                = 102;
  CT_OpSubtype_Data_Signer                = 103;
  CT_OpSubtype_Data_Receiver              = 104;

  CT_ClientAppVersion : String = {$IFDEF PRODUCTION}'4.1.2'{$ELSE}{$IFDEF TESTNET}'TESTNET 4.1.2'{$ELSE}{$ENDIF}{$ENDIF};

  CT_Discover_IPs = {$IFDEF PRODUCTION}'bpascal1.dynamic-dns.net;bpascal2.dynamic-dns.net;pascalcoin1.dynamic-dns.net;pascalcoin2.dynamic-dns.net;pascalcoin1.dns1.us;pascalcoin2.dns1.us;pascalcoin1.dns2.us;pascalcoin2.dns2.us'
                    {$ELSE}'pascaltestnet1.dynamic-dns.net;pascaltestnet2.dynamic-dns.net;pascaltestnet1.dns1.us;pascaltestnet2.dns1.us'{$ENDIF};

  CT_TRUE_FALSE : Array[Boolean] Of String = ('FALSE','TRUE');

  CT_MAX_0_fee_operations_per_block_by_miner = {$IFDEF PRODUCTION}30000{$ELSE}{$IFDEF TESTNET}3000{$ELSE}{$ENDIF}{$ENDIF};
  CT_MAX_Operations_per_block_by_miner =  {$IFDEF PRODUCTION}500000{$ELSE}{$IFDEF TESTNET}50000{$ELSE}{$ENDIF}{$ENDIF};

  CT_MAX_MultiOperation_Senders = 100;
  CT_MAX_MultiOperation_Receivers = 1000;
  CT_MAX_MultiOperation_Changers = 100;

  CT_DEFAULT_MaxSafeboxSnapshots = 10;

  CT_MOLINA  = 1;
  CT_MOLINA_DECIMAL = {$IFDEF FPC}Real(CT_MOLINA/1000.0);{$ELSE}0.0001;{$ENDIF}

  CT_ACTIVATE_RANDOMHASH_V4 = {$IFDEF ACTIVATE_RANDOMHASH_V4}True{$ELSE}False{$ENDIF};

  // App Params
  CT_PARAM_GridAccountsStream = 'GridAccountsStreamV2';
  CT_PARAM_GridAccountsPos = 'GridAccountsPos';
  CT_PARAM_DefaultFee = 'DefaultFee';
  CT_PARAM_InternetServerPort = 'InternetServerPort';
  {$IFDEF TESTNET}CT_PARAM_AutomaticMineWhenConnectedToNodes = 'AutomaticMineWhenConnectedToNodes';{$ENDIF}
  CT_PARAM_MinerPrivateKeyType = 'MinerPrivateKeyType';
  CT_PARAM_MinerPrivateKeySelectedPublicKey = 'MinerPrivateKeySelectedPublicKey';
  CT_PARAM_SaveLogFiles = 'SaveLogFiles';
  CT_PARAM_SaveDebugLogs = 'SaveDebugLogs';
  CT_PARAM_ShowLogs = 'ShowLogs';
  CT_PARAM_MinerName = 'MinerName';
  CT_PARAM_FirstTime = 'FirstTime';
  CT_PARAM_ShowModalMessages = 'ShowModalMessages';
  {$IFDEF TESTNET}CT_PARAM_MaxCPUs = 'MaxCPUs'; {$ENDIF} //deprecated
  CT_PARAM_PeerCache = 'PeerCache';
  CT_PARAM_TryToConnectOnlyWithThisFixedServers = 'TryToConnectOnlyWithFixedServers';
  CT_PARAM_JSONRPCMinerServerPort = 'JSONRPCMinerServerPort';
  CT_PARAM_JSONRPCMinerServerActive = 'JSONRPCMinerServerActive';
  CT_PARAM_JSONRPCEnabled = 'JSONRPCEnabled';
  CT_PARAM_JSONRPCAllowedIPs = 'JSONRPCAllowedIPs';



implementation

end.
